Odyssey compatibility (location, vehicle, and status)

Tkael opened the pull request:
- Adds 3 new vehicle types:
  - On Foot
  - MultiCrew
  - Taxi
- Updates the Status Monitor for Odyssey data and updates status documentation in Variables.md
- Prevents the Ship Monitor from interpreting suits as ships
- Replaces internal flags `inCQC` and `inMulticrew` with new internal flag `inTelepresence`, to better distinguish circumstances where we should not transmit data
- Updates the following events:
  - Touchdown (new properties)
  - Location (new properties)
  - Liftoff (new properties)
  - Commander Continued (new properties)
  - Died (revised properties, updated script)
  - Bounty awarded (updated script to handle non-ship targets)
  - Bond awarded (updated script to handle non-ship targets)
  - Fuel check (updated to prevent triggering when in a taxi or similar)
- Adds new events:
  - Embark
  - Disembark
- Updated {ShipName()} to provide a more appropriate response when in a taxi or multicrew scenario
- Set variables for `odyssey` and `horizons` game states, available to players in Cottle and VA

Tkael committed:
Add the `Odyssey` bool to the `LoadGame (CommanderContinued) event.

Tkael committed:
Update the `Touchdown` and `Liftoff` events
Note: I'm not sure how to interpret the new `OnStation` and `OnPlanet` bool, so I've commented those out for now.

Tkael committed:
Update `Location` event

Tkael committed:
Add unhandled events

Tkael committed:
Update `Embark` and `Disembark` events

Tkael committed:
Basic support for Odyssey legs detection
- Adds new events `Embark` and `Disembark`
- Adds new vehicle types `On Foot`, `Taxi`, and `Multicrew`
- Converts `inCQC` and `inCrew` bool to a single `inTelepresence` bool to help distinguish physical multicrew from telepresence multicrew and reduce redundancy

Add handling for ship monitor `Commander Continued` event with on-foot status.

Tidy using

Tkael committed:
Update the status monitor

Don't update our vehicle from status if we don't have an appropriate status flag.
Currently, status may have no value when we are not in a ship, srv, or fighter. If we find ourselves in that state, don't update the global vehicle string.

Tkael committed:
Do not pass events missing station names or add these to our local database.

richardbuckle commented on Utilities/Constants.cs:
```diff
@@ -25,6 +25,9 @@ public class Constants
         public static readonly string VEHICLE_SHIP = "Ship";
         public static readonly string VEHICLE_SRV = "SRV";
         public static readonly string VEHICLE_FIGHTER = "Fighter";
+        public static readonly string VEHICLE_LEGS = "On Foot";
```
Possibly `VEHICLE_SUIT = "Suit";`? I'm in two minds here.

richardbuckle replied:
Also "vehicle" has become a bit of a misnomer.

Tkael replied:
I don't think "Suit" is appropriate since it's also the status when you're wandering around a station concourse without a suit.

richardbuckle replied:
Au contraire, when wandering around the station concourse you are ALWAYS in a suit, even if it is just a vanilla flight suit.

richardbuckle replied:
You can also be in a Maverick, Artemis suit etc, but you are always in some sort of suit.

Tkael replied:
Fair, you've convinced me to be open minded but I still don't see the virtue of `Suit` over `On Foot`?

richardbuckle replied:
"Suit" is a noun like all the others, is all. Not a huge deal but it feels a bit more uniform.

Tkael replied:
I'm not sure. You make a good point but I think that people will still expect "On Foot" or something similar.

richardbuckle commented on EDDI/EDDI.cs:
```diff
@@ -1506,9 +1564,12 @@ private bool eventUndocked(UndockedEvent theEvent)
+            if (theEvent.playercontrolled && theEvent.latitude != null && theEvent.longitude != null)
             {
                 Environment = Constants.ENVIRONMENT_LANDED;
                 Vehicle = theEvent.playercontrolled ? Constants.VEHICLE_SHIP : Constants.VEHICLE_SRV;
```
We are in an `if (theEvent.playercontrolled && ...` block here. have we defeated the logic of this statement?

Tkael replied:
Concur. Reverted.

richardbuckle commented on EDDI/EDDI.cs:
```diff
@@ -1506,9 +1564,12 @@ private bool eventUndocked(UndockedEvent theEvent)
+            updateCurrentSystem(theEvent.systemname);
+            updateCurrentStellarBody(theEvent.bodyname, theEvent.systemname, theEvent.systemAddress);
+
+            // If playercontrolled, only pass on this event if our longitude and lattitude are set
```
typo "lattitude"

richardbuckle commented on EDDI/EDDI.cs:
```diff
@@ -1864,14 +1937,16 @@ private bool eventEnteredNormalSpace(EnteredNormalSpaceEvent theEvent)
         {
-            inCrew = false;
+            inTelepresence = false;
+            Vehicle = Constants.VEHICLE_SHIP;
```
I'm not sure about this. What if I entered multicrew when on foot or in an SRV?

Tkael replied:
The assumption is that the event only fires for telepresence multicrew. I'm not sure the event fires for physical multicrew. Does it?

Telepresence multicrew is not available when in an SRV (at least in the current live game)... there's no second seat in the SRV for a player two.

Is the option for telepresence multicrew available from on foot or from an SRV in Odyssey?

richardbuckle replied:
Good question, need to check.

richardbuckle replied:
I have confirmed that in Odyssey you can request multicrew from SRV and Suit, but I have not yet established what you return to.

richardbuckle commented on Events/DisembarkEvent.cs:
```diff
@@ -0,0 +1,87 @@
+        public bool frommulticrew { get; }
+
+        [PublicAPI]
+        public bool fromship => fromLocalId != null && !fromsrv && !fromtaxi && !frommulticrew;
```
I know it's a derived property but let's give it test coverage all the same. Similarly for the others below.

Tkael replied:
Please provide some raw snapshots of status.json at various instances you'd like to test? These aren't currently logged.

richardbuckle replied:
See below.

Tkael replied:
Done

richardbuckle commented on Events/LiftoffEvent.cs:
```diff
@@ -36,10 +52,20 @@ static LiftoffEvent()
+            this.systemAddress = systemAddress;
+            this.bodyname = body;
+            this.bodyId = bodyId;
+            this.onstation = onStation;
```
inconsistent capitalization

Tkael replied:
It's consistent with our standard practice of using lowercase for user facing variables and camel case for internal variables, but if you feel strongly about it then I can change it.

richardbuckle replied:
I was unaware of that convention.

richardbuckle commented on Events/TouchdownEvent.cs:
```diff
@@ -36,10 +52,20 @@ static TouchdownEvent()
+            this.systemAddress = systemAddress;
+            this.bodyname = body;
+            this.bodyId = bodyId;
+            this.onstation = onStation;
```
inconsistent capitalization

Tkael replied:
It's consistent with our standard practice of using lowercase for user facing variables and camel case for internal variables, but if you feel strongly about it then I can change it.

richardbuckle replied:
I wasn't aware of that convention. Let's leave it for now.

richardbuckle commented on StatusMonitor/Status.cs:
```diff
@@ -47,11 +47,32 @@ public enum Flags : uint
         // Variables set from status flags (when not signed in, this is set to '0')
         public string vehicle =>
-                ((flags & Flags.InSRV) != 0) ? Constants.VEHICLE_SRV
+                ((flags & Flags.InMainShip) != 0) ? Constants.VEHICLE_SHIP
```
This gets a bit complex so let's add test coverage.

Tkael replied:
Please provide some raw snapshots of status.json at various instances you'd like to test? These aren't currently logged.

richardbuckle replied:
provided below

Tkael replied:
Added.

richardbuckle committed:
typo

richardbuckle committed:
Temp logging for Odyssey

richardbuckle commented:
Adding play session journals and logs
[JournalAlpha.210430234357.01.log](https://github.com/EDCD/EDDI/files/6408557/JournalAlpha.210430234357.01.log)
[odyssey test.txt](https://github.com/EDCD/EDDI/files/6408558/odyssey.test.txt)
[eddi.log](https://github.com/EDCD/EDDI/files/6408559/eddi.log)

richardbuckle commented:
I will expand odyssey test.txt here:

--------------

2021-04-30T21:30:08 [Info] EDDI:eventCommanderContinued Commander Continued: vehicle is On Foot
Launched on foot in station: Status.vehicle is "" despite the above.

2021-04-30T21:30:41 [Warning] JournalMonitor:ParseJournalEntry Value cannot be null.
Parameter name: Expected value for Taxi not present

2021-04-30T21:30:43 [Warning] JournalMonitor:ParseJournalEntry EDDI has no handler for event type 'BackPackMaterials'
The above occurs a lot.

Took a taxi
2021-04-30T21:53:37 [Info] EDDI:eventEmbark Embarked to Taxi
2021-04-30T21:53:41 [Info] EDDI:eventMusic Music event for vehicle Taxi
2021-04-30T21:53:42 [Info] StatusMonitor:handleStatus Status changed vehicle from  to Ship

Relogged while in taxi
2021-04-30T22:00:06 [Warning] JournalMonitor:ParseJournalEntry Value cannot be null.
Eddi said "entering close quarters combat" (yes really)
Taxi low waked
EDDI said "Fair Kaleen has entered Supercruise"
That is the name of the last ship I was in.

While in taxi lots of:
2021-04-30T22:00:42 [Info] EDDI:eventMusic Music event for vehicle Ship
I don't think we should rely on the music events any more, too fragile.

2021-04-30T22:06:12 [Warning] JournalMonitor:ParseJournalEntry Unparseable value for SystemAddress

2021-04-30T22:06:23 [Warning] JournalMonitor:ParseJournalEntry EDDI has no handler for event type 'TransferMicroResources'

Relogged while on foot on planet
2021-04-30T22:06:26 [Info] EDDI:eventMusic Music event for vehicle On Foot
2021-04-30T22:06:27 [Info] StatusMonitor:handleStatus Status changed vehicle from Ship to On Foot
so both GOOD

Boarded taxi from planet, EDDI said: Full control re-established.
2021-04-30T22:16:57 [Info] StatusMonitor:handleStatus Status changed vehicle from On Foot to Ship
2021-04-30T22:17:08 [Info] EDDI:eventMusic Music event for vehicle Ship
2021-04-30T22:17:27 [Info] EDDI:eventMusic Music event for vehicle Ship
^^^^ i.e. not Taxi

2021-04-30T22:00:06 [Warning] JournalMonitor:ParseJournalEntry Value cannot be null.
Parameter name: Expected value for InSRV not present

Relog in own ship
No issues

Tkael committed:
Remove unnecessary (and less reliable) music event for vehicle detection.

Tkael committed:
Revert Touchdown event logic

Tkael committed:
Add (empty) default scripts for `Embark` and `Disembark` events.

Tkael committed:
The `Taxi`, `Multicrew`, `InSRV`, and `OnFoot` properties of the `Location` event are optional bool, only written when true.

Tkael committed:
Fix typo in `BackPackMaterials` unhandled event

Tkael committed:
Moved status logging for vehicle switch so that it can't be triggered by refueling.

Tkael committed:
Add raw status to the status object for debugging purposes.

Tkael committed:
`Docked` at `SurfaceStation` currently returns an abbreviated event like
```json
{ "timestamp":"2021-04-30T22:06:14Z", "event":"Docked", "StationName":"", "StationType":"SurfaceStation", "Taxi":true, "Multicrew":false, "DistFromStarLS":3396.001838 }
```
Neither SystemAddress nor MarketID are present. We need to handle this gracefully.

Tkael committed:
Add to unhandled events the undocumented event `TransferMicroResources`.

Tkael committed:
Add new station service `socialspace`

Tkael commented:
@richardbuckle 
> "Eddi said "entering close quarters combat" (yes really)"...

We generate a derived event for entering CQC when we see `Loadgame` journal event (a Commander continued EDDI event) without a `ShipID`. This generally works for Odyssey because your suit has a value but it may not work if you re-log while when in a taxi, while on foot in a concourse, or when in physical multicrew? 

I think we need to further explore this. We may need to rethink CQC detection given the changes with Odyssey.

I don't see an event for re-logging while in a Taxi in the log you provided to use a comparison (the journal log timestamps don't match eddi.log timestamps either... you may have uploaded a journal file from a different play session?)

Please provide `Loadgame` events from re-logging when in a taxi and from re-logging in physical multicrew?

EDIT: Sure enough, the taxi case has no ShipID. Handled.

Tkael commented:
> Taxi low waked
> EDDI said "Fair Kaleen has entered Supercruise"
> That is the name of the last ship I was in.

There are two ways that we could handle this. We could either
1. Suppress certain events from triggering while in a taxi OR
2. Allow events to trigger, but replace references to "your ship" with references to "your Apex taxi service".

Thoughts?

Affects: `FSD engaged`, `Entered supercruise`, `Docked`... any other events?

Tkael commented:
> Boarded taxi from planet, EDDI said: Full control re-established.
> 2021-04-30T22:16:57 [Info] StatusMonitor:handleStatus Status changed vehicle from On Foot to Ship
> 2021-04-30T22:17:08 [Info] EDDI:eventMusic Music event for vehicle Ship
> 2021-04-30T22:17:27 [Info] EDDI:eventMusic Music event for vehicle Ship
> ^^^^ i.e. not Taxi

The `Undocked` event currently sets our vehicle to `Ship`. Apparently the event also triggers while in Taxi. Same comments as https://github.com/EDCD/EDDI/pull/2163#issuecomment-830492933... 
I'll remove the vehicle switching based on the `Undocked` event, but from a user experience perspective how do you think we ought to handle these events while in a taxi?

Tkael committed:
We can't guarantee that we'll be in our ship when we dock (might be in multicrew or a taxi)

Tkael committed:
Only set vehicle to `Ship` for `Location` events where we are docked but not in a taxi or multicrew.

Tkael committed:
The `Undocked` event may trigger when our vehicle is `taxi` or `multicrew`, we can't assume `ship` is accurate.

Tkael commented:
Can the `Touchdown` event fire with the `playercontrolled` property false when you touch down on a surface from a taxi? Or do does the taxi always dock rather than touch down?

Context: We currently set vehicle info from the Touchdown event as follows:
```cs
                Vehicle = theEvent.playercontrolled ? Constants.VEHICLE_SHIP : Constants.VEHICLE_SRV;
```

Tkael commented:
Similar question for the `Liftoff` event. If it can fire with the `playercontrolled` property false when you lift off from a surface in taxi then we may set variables incorrectly.

Context: We currently set vehicle info from the LiftOff event as follows:
```cs
            if (theEvent.playercontrolled)
            {
                Vehicle = Constants.VEHICLE_SHIP;
            }
            else
            {
                Vehicle = Constants.VEHICLE_SRV;
            }
```

Tkael committed:
Reduce assumptions about when we're in our ship.

Tkael commented:
In physical multicrew, can you dock someone else's SRV at their ship? Launch some else's SRV? Get destroyed in some else's SRV?
Same question for fighters also.

richardbuckle commented:
This is the log file we need 
[JournalAlpha.210430222855.01.log](https://github.com/EDCD/EDDI/files/6410217/JournalAlpha.210430222855.01.log)

richardbuckle commented:
Line 161 for the relog in suit.
Line 288 for the relog in taxi.
Also, I have realised that "Fair Kaleen" is NOT the ship I was last in; that would be my Krait Mk II "Convergence". "Fair Kaleen" is my Cutter, which I have not flown in the beta at all.

richardbuckle commented:
Line 291 for the Location in taxi and 358 for low wake in taxi.

richardbuckle commented:
TransferMicroResources is probably when you transfer on-foot items between suit and ship.

richardbuckle commented:
Confirmed: yes one can initiiate telepresence multicrew from SRV and on foot. Didn't have any friends to try it with.
Similarly cannot really try physical multicrew without a willing test partner.

2021-05-01T22:11:17 Confirmed it is possible to disembark to SRV, then go on foot, then re-embark to ship from on foot and leave your SRV behind. If you lo-wake or hi-wake out then the SRV will be destroyed. 

The game *does* warn you but only when you engage the FSD, and only via a silent onscreen message. If you don't disengage the FSD then it's too late. *This could be a big deal for explorers as an SRV cannot be synthesized.*

**Opportunity: we could help CMDRs HUGELY by warning them when they embark on foot with an SRV unaccounted for.**

Similarly it is possible to disembark, summon a taxi and leave your ship behind. However in this case the ship persists and can be summoned back to port for a small fee. This, I feel is a minor thing with no great consequence. Explorers don't summon taxis.

Re taxis, we have a couple more erroneous transitions to `Ship` to squash: 
	2021-05-01T22:29:03 [Info] EDDI:eventEmbark Embarked to Taxi
	2021-05-01T22:29:04 [Info] StatusMonitor:handleStatus Status changed vehicle from On Foot to On Foot ...
	2021-05-01T22:29:07 [Info] StatusMonitor:handleStatus Status changed vehicle from On Foot to Ship: ...
The first is good but the last two are wrong. 
**Suggestion: when in Taxi, ignore everything until eventDisembark.**
This is starting to look like we want a finite state machine.

`TransferMicroResources` seems bugged, see Discord `EDCD/#odyssey-alpha`. 
Only written when you transfer when on foot and ready to board ship or SRV. 
Not written when using the Storage tab in the right-hand panel, moreover those choices appear to be pending disembarkation and are discarded if you relog before disembarking. 
I recommend holding off on any attempt to develop based on this until FD stabilise it.

{ "timestamp":"2021-05-01T21:36:45Z", "event":"TransferMicroResources", ...}
{ "timestamp":"2021-05-01T21:36:46Z", "event":"TransferMicroResources", ...}

More about taxis.

2021-05-01T22:27:45Z summoned taxi at random point on planet surface. As you see we get 
{ "timestamp":"2021-05-01T22:27:45Z", "event":"BookTaxi", "Cost":100, "DestinationSystem":"Nervi", ...}
{ "timestamp":"2021-05-01T22:29:06Z", "event":"Embark", "SRV":false, "Taxi":true, ...}
**no Liftoff event**
{ "timestamp":"2021-05-01T22:29:50Z", "event":"SupercruiseEntry", "Taxi":true, ...}

So no LiftOff event in Taxi, I'll wager no Touchdown either.

Abandoned a conflict and booked a taxi. As above, was getting 
CRASH TEST DUMMY has entered supercruise.
CRASH TEST DUMMY has left supercruise near Coriolis Starport, Savitskaya Vision.
Fixing the above should fix this.
{ "timestamp":"2021-05-01T23:20:57Z", "event":"SupercruiseEntry", "Taxi":true, "Multicrew":false, "StarSystem":"Nervi", "SystemAddress":2518721481067 }

[JournalAlpha.210502000445.01 - Copy.log](https://github.com/EDCD/EDDI/files/6410528/JournalAlpha.210502000445.01.-.Copy.log)
[JournalAlpha.210501223326.01 - Copy.log](https://github.com/EDCD/EDDI/files/6410529/JournalAlpha.210501223326.01.-.Copy.log)
[JournalAlpha.210501215817.01 - Copy.log](https://github.com/EDCD/EDDI/files/6410531/JournalAlpha.210501215817.01.-.Copy.log)

richardbuckle commented:
Status snapshots, named by status:

[Odyssey status.zip](https://github.com/EDCD/EDDI/files/6410537/Odyssey.status.zip)

richardbuckle commented:
> e

Tkael committed:
Distinguish logging into CQC from logging in while in an Apex taxi. Add unit tests.

Tkael committed:
Update `Liftoff` and `Touchdown` samples, add undocumented taxi and multicrew properties.

richardbuckle commented:
By safely re-enabling the Inara and EDSM responders for beta (by disabling their send functions) I have established that when in a taxi, `InaraResponder.handleJumpedEvent()` and `InaraResponder.handleDockedEvent()` are indeed called, and thus will need adjustment.

Events:

```json
{ "timestamp":"2021-05-03T21:54:20Z", "event":"FSDJump", "Taxi":true, "Multicrew":false, "StarSystem":"Sumod", "SystemAddress":3961847269739, "StarPos":[-60.37500,-77.81250,45.62500], "SystemAllegiance":"Independent", "SystemEconomy":"$economy_Industrial;", "SystemEconomy_Localised":"Industrial", "SystemSecondEconomy":"$economy_Refinery;", "SystemSecondEconomy_Localised":"Refinery", "SystemGovernment":"$government_Corporate;", "SystemGovernment_Localised":"Corporate", "SystemSecurity":"$SYSTEM_SECURITY_high;", "SystemSecurity_Localised":"High Security", "Population":40786853, "Body":"Sumod A", "BodyID":2, "BodyType":"Star", "JumpDist":17.741, "FuelUsed":1.157341, "FuelLevel":8.000000, "Factions":[ { "Name":"Future of Sumod", "FactionState":"None", "Government":"Democracy", "Influence":0.071573, "Allegiance":"Independent", "Happiness":"$Faction_HappinessBand2;", "Happiness_Localised":"Happy", "MyReputation":0.000000, "PendingStates":[ { "State":"CivilWar", "Trend":0 } ] }, { "Name":"Sumod United Services", "FactionState":"None", "Government":"Corporate", "Influence":0.071573, "Allegiance":"Federation", "Happiness":"$Faction_HappinessBand2;", "Happiness_Localised":"Happy", "MyReputation":0.000000, "PendingStates":[ { "State":"CivilWar", "Trend":0 } ] }, { "Name":"Noblemen of Wadjarata", "FactionState":"None", "Government":"Feudal", "Influence":0.225806, "Allegiance":"Empire", "Happiness":"$Faction_HappinessBand2;", "Happiness_Localised":"Happy", "MyReputation":0.000000 }, { "Name":"Sumod Empire Group", "FactionState":"None", "Government":"Patronage", "Influence":0.057460, "Allegiance":"Empire", "Happiness":"$Faction_HappinessBand2;", "Happiness_Localised":"Happy", "MyReputation":0.000000 }, { "Name":"Sumod Gang", "FactionState":"None", "Government":"Anarchy", "Influence":0.021169, "Allegiance":"Independent", "Happiness":"$Faction_HappinessBand2;", "Happiness_Localised":"Happy", "MyReputation":0.000000 }, { "Name":"Corvus Corporation", "FactionState":"None", "Government":"Corporate", "Influence":0.552419, "Allegiance":"Independent", "Happiness":"$Faction_HappinessBand2;", "Happiness_Localised":"Happy", "MyReputation":0.000000 } ], "SystemFaction":{ "Name":"Corvus Corporation" }, "Conflicts":[ { "WarType":"civilwar", "Status":"pending", "Faction1":{ "Name":"Future of Sumod", "Stake":"Mallett Orbital", "WonDays":0 }, "Faction2":{ "Name":"Sumod United Services", "Stake":"Mather's Folly", "WonDays":0 } } ] }

{ "timestamp":"2021-05-03T21:56:46Z", "event":"Docked", "StationName":"Sharp Dock", "StationType":"Coriolis", "Taxi":true, "Multicrew":false, "StarSystem":"Sumod", "SystemAddress":3961847269739, "MarketID":3223952128, "StationFaction":{ "Name":"Corvus Corporation" }, "StationGovernment":"$government_Corporate;", "StationGovernment_Localised":"Corporate", "StationServices":[ "dock", "autodock", "commodities", "contacts", "exploration", "missions", "outfitting", "crewlounge", "rearm", "refuel", "repair", "shipyard", "tuning", "engineer", "missionsgenerated", "flightcontroller", "stationoperations", "powerplay", "stationMenu", "carriervendor", "livery", "modulepacks", "socialspace" ], "StationEconomy":"$economy_Industrial;", "StationEconomy_Localised":"Industrial", "StationEconomies":[ { "Name":"$economy_Industrial;", "Name_Localised":"Industrial", "Proportion":0.800000 }, { "Name":"$economy_Refinery;", "Name_Localised":"Refinery", "Proportion":0.200000 } ], "DistFromStarLS":7.002534 }

```

Tkael committed:
Add updated vehicle and location properties to `FSDJump`, `EnteredSupercruise`, and `EnteredNormalSpace` events.

Tkael committed:
Update variables.md description of vehicles

Tkael committed:
Add new definitions for `Suit` and `NpcSuitLoadout`. Update the `Died` event to better describe the context of the player's death.

Tkael committed:
Add civilian workers to our definitions.

Tkael committed:
Update the `Bounty awarded` event to handle on-foot targets.

Tkael committed:
Update the `ShipName()` function to return appropriate results when invoked from a taxi or multicrew.

Tkael committed:
Remember our former vehicle when joining and returning from telepresence multicrew.

Tkael committed:
Only report ship data to Inara when we jump or dock in our own ship.

Tkael committed:
Fix erroneous vehicle status when in a taxi

Tkael committed:
Touch up bond awarded script to permit bonds from destroying non-ship targets.

Tkael committed:
Prefer `Your transport` over `Your taxi` to better capture both taxi and dropship transport.

Tkael committed:
Note Howard's clarification of `InMainShip` meaning.

Tkael committed:
Add new on foot crimes

Tkael committed:
Add Odyssey docked status test

Tkael committed:
Add "DropshipDeploy" to unhandled events

Tkael committed:
Report Horizons and Odyssey game states to EDDN, keeping EDDN data parsing as direct and as simple as possible. Sanity check for empty station names.

Tkael committed:
Further enhance Odyssey game state tracking

Tkael committed:
Make horizons and odyssey game states available to players.

Tkael committed:
Update the `Fuel check` script to prevent inadvertant triggering when in a taxi. or similar.

Tkael committed:
Fix a parsing error that can occur when `SystemAddress` is null in a `Docked` event.

Tkael committed:
More gracefully check for matching killer and bounty ednames from definitions.

Tkael committed:
Add a status unit test when docked in a dropship

Tkael committed:
Add a status unit test when in a fighter

Tkael committed:
Add a status unit test when in supercruise in a taxi

Tkael committed:
Add a status unit test when landed in your ship

Tkael committed:
Add a status unit test when in an SRV with the ship dismissed

Tkael committed:
Add a status unit test when on foot on a planet

Tkael committed:
Add a status unit test when on foot on a orbital station, in a social space

Tkael committed:
Touch up logic for switching vehicle from status.

Tkael committed:
Improve our tests for matching objects with a new EDNameExists() method.

Tkael committed:
Add `Embark` unit tests. bodyId to int?

Tkael committed:
Add `Disembark` unit tests

Tkael committed:
Prevent ship damage from affecting our voice when we're testing scripts in the speech responder, or when we're on foot in a suit.

Tkael committed:
Mix mismatched systemAddress parsing and type, prefer "var".

Tkael committed:
Fix an exception that can occur if the commander's Frontier API capabilities / contexts list does not include Odyssey (or any other similar item that we check for).

Tkael committed:
Tidy up vehicle logging, move `onplanet` and `onstation` from user facing to hidden values in the `Liftoff` and `Touchdown` events - they do not appear to add any value.
